creme/linear_model/glm.py

```diff
@@ -55,37 +69,31 @@ def _fit(self, x, y, w, get_grad):
         self.intercept -= self.intercept_lr.get(self.optimizer.n_iterations) * loss_gradient
 
         # Update the weights
-        self.weights = self.optimizer.update_after_pred(w=self.weights, g=gradient)
+        self.optimizer.update_after_pred(w=self._weights, g=gradient)
 
         return self
 
     # Single instance methods
 
     def _raw_dot_one(self, x: dict) -> float:
-        return utils.math.dot(self.weights, x) + self.intercept
+        return self._weights @ utils.VectorDict(x) + self.intercept
 
     def _eval_gradient_one(self, x: dict, y: float, w: float) -> (dict, float):
 
         loss_gradient = self.loss.gradient(y_true=y, y_pred=self._raw_dot_one(x))
         loss_gradient *= w
-        loss_gradient = np.clip(loss_gradient, -self.clip_gradient, self.clip_gradient)
+        loss_gradient = np.clip(loss_gradient, -self.clip_gradient, self.clip_gradient).item()
 
-        return (
-            {
-                i: xi * loss_gradient + 2. * self.l2 * self.weights.get(i, 0)
-                for i, xi in x.items()
-            },
-            loss_gradient
-        )
+        return (loss_gradient * utils.VectorDict(x)  + 2. * self.l2 * self._weights, loss_gradient)
```

Comment:
No need for parenthesis. 

I assume that we could write a bespoke Cython function to do this in one single loop?

Reply:
Yes could do that, should I add to this PR? As mentioned in another comment there may be several points to optimize, would need to profile to check where to focus.